tony-core/src/main/java/com/linkedin/tony/ApplicationMaster.java

```diff
@@ -402,25 +405,34 @@ private boolean prepare() {
 
     RegisterApplicationMasterResponse response;
     try {
-      response = amRMClient.registerApplicationMaster(hostname, amPort, null);
-      amHostPort = hostname + ":" + amPort;
-      LOG.info("RPC server running at: " + amHostPort);
-      if (secureMode) {
-        ClientToAMTokenIdentifier identifier =
-            new ClientToAMTokenIdentifier(appAttemptID, user);
-        byte[] secret = response.getClientToAMTokenMasterKey().array();
-        ClientToAMTokenSecretManager secretManager = new ClientToAMTokenSecretManager(appAttemptID, secret);
-        Token<? extends TokenIdentifier> token = new Token<>(identifier, secretManager);
-        token.setService(new Text(amHostPort));
-        rpcServer.setSecretManager(secretManager);
-        UserGroupInformation.getCurrentUser().addToken(token);
-        setupContainerCredentials();
-      }
-    } catch (Exception e) {
+      response = amRMClient.registerApplicationMaster(amHostname, amPort, null);
+      amHostPort = amHostname + ":" + amPort;
+    } catch (YarnException e) {
       LOG.error("Exception while preparing AM", e);
       return false;
     }
 
+    if (secureMode) {
```

Comment:
Add a try catch block around this, currently if a cluster is not secure but I don't specify insecure mode, there will be a vague exception thrown out. We should explicitly tell the user the mistake

Reply:
I let any `IOException` propagate and crash the AM. There should be a stack trace.

To really debug this when I was developing, I had to enable debug logging:
```
org.apache.log4j.Logger logger4j = org.apache.log4j.Logger.getRootLogger();
logger4j.setLevel(org.apache.log4j.Level.toLevel("DEBUG"));
```

or change the logging level in the tony-cli `log4j.properties`.

Reply:
The stack trace is like - failed to get RM token. We should add something average people could understand..

Reply:
Hmm, I felt an error message didn't help much when developing; I had to enable DEBUG logging to really understand what was going on. That said, if I encounter issues with this code in the future and this part of the code is the culprit, I'll add some more logging then.